### ISSUE TITLE
Multistage Dockerfile does not work with WORKDIR

### ISSUE THREAD

**Issue by Miouge1** [moby/moby#36643]
**Description**

When using multistage Dockerfile WORKDIR is not working as expected:
**Steps to reproduce the issue:**
```
FROM alpine as base
WORKDIR /foo
RUN echo test > test.txt

FROM alpine
WORKDIR /bar
COPY --from=base test.txt .
```

**Describe the results you received:**
```
COPY failed: stat /var/lib/docker/overlay2/..../merged/test.txt: no such file or directory
```
**Describe the results you expected:**
A succesful build

**Additional information you deem important (e.g. issue happens only occasionally):**
When using `--from` I expect to be able to use paths relative to the WORKDIR of the origin container
**Output of `docker version`:**

```
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:03:51 2017
 OS/Arch:	darwin/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:12:29 2017
  OS/Arch:	linux/amd64
  Experimental:	true
```

**Output of `docker info`:**

```
docker info
Containers: 88
 Running: 3
 Paused: 0
 Stopped: 85
Images: 140
Server Version: 17.12.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.60-linuxkit-aufs
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: YJ2S:B6MH:6A5I:THF6:PQAE:ZJ3Y:EQ54:ISR3:Z4R6:CP6Q:VEBG:ANC5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 33
 Goroutines: 56
 System Time: 2018-03-20T17:45:55.558525421Z
 EventsListeners: 3
HTTP Proxy: docker.for.mac.http.internal:3128
HTTPS Proxy: docker.for.mac.http.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
I think this is the expected behaviour; `COPY` takes a full path where to copy from, and does not take the `WORKDIR` into account.

This is similar to 

```Dockerfile
FROM alpine as base
WORKDIR /foo
RUN echo test > test.txt
```

```bash
docker build --no-cache -t foobar

docker create --name foocontainer foobar

docker cp foocontainer:test.txt .
Error: No such container:path: foocontainer:test.txt

docker cp foocontainer:/foo/test.txt .
```


ping @tonistiigi to confirm, but I think this is the expected behaviour

**Comment by Miouge1** [comment]
Since `<dst>` is relative to the current `WORKDIR`, a user can expect `<src>` to also be relative to `WORKDIR` when using `--from`.


As far as I have tested `COPY --from=something <src>` allows for relative paths, but they are relative to /, for example:

```
FROM alpine as base
WORKDIR /foo
RUN echo test > test.txt

FROM alpine
WORKDIR /bar
COPY --from=base foo/test.txt .
```

So, maybe I can re-classify this as a feature request then: when using the `--from` flag, `COPY` should interpret `<src>` as relative to the `WORKDIR` of that container

Impact should be minimal:
- `COPY --from /foo/test.txt .`: would be unchanged
- `COPY --from foo/test.txt .`: would not work anymore, but anyway what is the point of a path relative to /?
- `COPY --from test.txt .`: would work

**Comment by tonistiigi** [comment]
I remember the questions if these paths should be relative or not was brought up in the original PR. I wouldn't mind putting them in the context of workdir of source stage but this may break some existing users who don't use it currently without the slash.

**Comment by Miouge1** [comment]
Is the behavior empiric or documented somewhere? If it's not documented then it shouldn't be too many users impacted

AFAIK https://docs.docker.com/engine/reference/builder/#copy doesn't cover the case of relative `<src>` together with `--from`.

**Comment by jer-k** [comment]
Sounds like this was never implemented then? I'm trying to create a multistage build and copy in the contents from a previous image where the `WORKDIR` is not known to me. By doing a `COPY --from=base_image . .`, I am receiving the code that is in the `WORKDIR` but I have no way of knowing where that code is. 

I went down the path of trying to write the current working directory into a file, copy that file to the second stage and use that during the copy command, something along the lines of `COPY --from=base_image $(cat ./base_image_workdir.txt) .` to no avail. It could be that I just need different syntax in the copy command to make it work, but I couldn't find any information related to that.

Is there any solution to understanding what the `WORKDIR` from a previous stage is?

**Comment by thaJeztah** [comment]
> something along the lines of `COPY --from=base_image $(cat ./base_image_workdir.txt) .`

That won't work indeed, because `COPY` instructions don't run in a shell

> Is there any solution to understanding what the `WORKDIR` from a previous stage is?
 
No, I don't think there is; best option is to have a predictable path in the stage (or image) you're copying from.


Note that `COPY` (without `--from`) never used `WORKDIR` for the `<src>` (`<src>` has always been relative to the "root" of the build-context), so from that perspective it's not "unexpected".

I can see use-cases for this though, but also share @tonistiigi's concern https://github.com/moby/moby/issues/36643#issuecomment-374735789

**Comment by jer-k** [comment]
Thanks for the update. The issue I'm facing is that I'm building images for other people so I have no control over setting up a predictable path unfortunately.

**Comment by thaJeztah** [comment]
> The issue I'm facing is that I'm building images for other people so I have no control over setting up a predictable path unfortunately.

Are you building those images, or are you `COPY --from..` from an image they built? Because in the first case, wouldn't the Dockerfile show what the `WORKDIR` is of the stage (so you could adjust your `<src>` accordingly?)

**Comment by jer-k** [comment]
Yes I am building them and copying. Its very basic once I've built their build and I have my wrapper build.

```
FROM their_build as builder

FROM my_wrapper_build

COPY --from=builder . .

CMD ["./script.sh"]
```

Are you suggesting parsing their raw `Dockerfile` to find the `WORKDIR` or is there something else I could do to get the value? I was thinking of writing a parser for their `Dockerfile` to understand what to expect.

**Comment by thaJeztah** [comment]
> Are you suggesting parsing their raw Dockerfile to find the WORKDIR or is there something else I could do to get the value? I was thinking of writing a parser for their Dockerfile to understand what to expect.

No, I wouldn't parse the Dockerfile, because that wouldn't give you the `WORKDIR` if it was set in a parent image they use. Also, `WORKDIR` could use `ENV` or `ARG`'s in it's path, so that would require you to handle those as well.
From your description, what you're interested in is the workdir in their final image, and that can be found in the image config;


```bash
docker build -t their_build -<<EOF
FROM busybox
WORKDIR /foo/bar/
EOF
```

Using `docker inspect` you can find the workdir;

```bash
docker image inspect --format '{{.Config.WorkingDir }}' their_build:latest
/foo/bar/

export BUILD_WORKDIR="$(docker image inspect --format '{{.Config.WorkingDir }}' their_build:latest)"
```

And you could pass that information to your second build as a build-arg. That is of course assuming that;

- your customer's image has all the files you need stored in the `WORKDIR` of the image (and you don't need anything from out side of that)
- your customer's image has a `WORKDIR` set (the default is `""` or `"/"`, and you likely don't want to `COPY` the whole rootfs of their image. You could check for that situation in your Dockerfile though. Something like;


```bash
export BUILD_WORKDIR="$(docker image inspect --format '{{.Config.WorkingDir }}' their_build:latest)"

docker build --build-arg BUILD_WORKDIR=$BUILD_WORKDIR -<<EOF
FROM their_build as builder

FROM my_wrapper_build
ARG BUILD_WORKDIR
RUN if [ "${BUILD_WORKDIR}" = "" ] || [ "${BUILD_WORKDIR}" = "/" ]; then echo "Invalid WORKDIR" && exit 1; fi
COPY --from=builder $BUILD_WORKDIR/. .

CMD ["./script.sh"]
EOF
```

**Comment by jer-k** [comment]
Thank you so much for taking the time to write this up, I really appreciate it!

I was thinking through the parsing of the Dockerfile after I had posted my comment and realized that there are many edge cases, especially the one you mentioned with an inherited `WORKDIR`. I did not know that you could find the `WORKDIR` from `inspect` that is really cool and I think it will solve my issue. Thank you again!

**Comment by tonistiigi** [comment]
If we can think of some microformat to signify that the path is meant to be relative we could theoretically add this but we're way past the point where we could just change the behavior. I think even changing paths that begin with `./` would be risky. The original thinking behind leaving the path as only absolute was probably that when you are using `COPY` without `--from` the path doesn't need to start with `/` either but is still always absolute to the context.

**Comment by winmillwill** [comment]
I would happily adopt `COPY --from=foo --workdir bar` to get `bar` from the `WORKDIR` of stage `foo`, if that's feasible to implement.
